File: 345-reverse-vowels-of-a-string/main.go
func reverseVowels(s string) string {
    
    vowels := "aeiouAEIOU"
	vowelSet := make(map[rune]bool)
	for _, v := range vowels {
		vowelSet[v] = true
	}

	runes := []rune(s)
	left, right := 0, len(runes)-1

	for left < right {
		if vowelSet[runes[left]] && vowelSet[runes[right]] {
			runes[left], runes[right] = runes[right], runes[left]
			left++
			right--
		}
		if !vowelSet[runes[left]] {
			left++
		}
		if !vowelSet[runes[right]] {
			right--
		}
	}

	return string(runes)
}

File: 345-reverse-vowels-of-a-string/main.rs
use std::collections::HashSet;

impl Solution {
    pub fn reverse_vowels(s: String) -> String {

        let vowels: &str = "aeiouAEIOU";
        let vowel_set: HashSet<char> = vowels.chars().collect();

        let mut chars: Vec<char> = s.chars().collect();
        let (mut left, mut right) = (0, chars.len() - 1);

        while left < right {
            if vowel_set.contains(&chars[left]) && vowel_set.contains(&chars[right]) {
                chars.swap(left, right);
                left += 1;
                right -= 1;
            }
            if !vowel_set.contains(&chars[left]) {
                left += 1;
            }
            if !vowel_set.contains(&chars[right]) {
                right -= 1;
            }
        }

        chars.iter().collect()
    }
}
